### PR TITLE
A new try on algebra homomorphisms

### PR DESCRIPTION
**Pull request by raulepure**
In this PR I try to unify algebra homomorphisms for MPoly and MPolyQuo.
Standard properties as composition, codomain, domain, kernel, preimage
of ideals, checks for injectivity, surjectivity and bijectivity.
I adapted existing code to work with this.

### CODE REVIEW

**Comment by fieker** [comment]
ERROR: LoadError: LoadError: UndefVarError: MPolyQuoHom not defined

in the doc tests. I assume the docu has to be updated as well?

**Comment by raulepure** [comment]
I forgot to update this one. Will be done today.

File: src/Rings/affine-algebra-homs.jl [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+###############################################################################
+
+#Compute the type of the base_ring of the underlying poly ring
+function _type_helper(R)
+   if isdefined(R, :I)
+      return typeof(base_ring(R.R))
+   else
+      return typeof(base_ring(R))
+   end
+end
```
Comment (thofma) [inline src/Rings/affine-algebra-homs.jl:150]: I don't think this is a good idea.

Comment (raulepure) [reply src/Rings/affine-algebra-homs.jl:150]: What do you mean? 
I just need to know over which field we consider our ring as an algebra and base_ring returns the underlying polynomial ring if it is a MPolyQuo.

**Comment by thofma** [comment]
Why not parametrize the map using the type of the domain? All these type computations based on `isdefined(R, ...)` shenanigans are kind of bad.

**Comment by raulepure** [comment]
I tried parametrizing and it caused trouble in the normalization code since MPolyQuoElem{Singular.spoly} is somehow allowed and this caused trouble. 
Furthermore, the conversion of elements did not work by simply doing Q(x) so I had to separate.
I agree that these things are not great, for the moment it is just supposed to work.

File: src/Rings/affine-algebra-homs.jl (hunk 1 above)
Comment (fingolfin) [reply src/Rings/affine-algebra-homs.jl:150]: I am also confused what this does: you check for the presence of a field `R.I` and if it is there, access a different field `R.R`? And which type(s) is `R` even allowed to have? Looking at the call in `AlgebraHomomorphism`, it seems `D` is either an `MPolyRing{T}` or a `MPolyQuo`. In which of those is `R.I` present? My guess/hope is it is present exactly in `MPolyQuo`? If so, then far better would to dispatch on that type... 

In any case, if it works right now, that's OK, and we can "improve" it later on

**Pull request merged**